### PR TITLE
Feature/#107 testing strategy cypress

### LINKED ISSUE #107: Establishing testing strategy [Component Library]
**A testing strategy should be defined for the components**

## Details
Currently, no decisions have been made about testing. Now that we start to style the components, we should decide how we test (what kind of tests, which frameworks and a targeted coverage) them to ensure it remains functional after a change has been made and to ensure a newly added piece of functionality works as expected.

If the functionality that needs to be tested at the moment we start working on this story is still limited, apply the testing strategy to all current code where it's deemed necessary. If the amount of untested code is significant, a separate story should be made to create the tests.

## Acceptance criteria
- [x] Desired types of tests are determined and discussed (smoke, functional, unit)
- [x] Research how to test a11y.
- [x] Suggestions or decisions for frameworks are determined
- [x] Write example test
- [x] Testing guidelines are documented in README.md

### PR DESCRIPTION
**Pull request by DanielvanVliet**
Fixes #107

In the component library we should test each component individually using multiple type of tests.

**Snapshot testing**
The first type of test we implement is snapshot testing.
This is useful to track any (possibly unexpected) changes in the component.

See https://jestjs.io/docs/snapshot-testing for more details.

**Visual regression testing**
This is similar to snapshot testing, except that image snapshots of rendered components are compared. I personally think that text-based snapshots suffice for now.
Another idea is to compare the rendered components with images of the Figma design, but this will likely involve having to manually create such images. This requires too much maintenance in the long run when designs change and is not ideal.

Since we already have snapshot testing and due to the additional test maintenance we will not currently use  this.

**Accessibility**
Testing accessibility with automated tools is far from perfect, different people claim that 50% - 70% of issues are found. This means that there is always some manual review necessary to ensure all issues are found. The automated tests for accessibility seem to function as static linters that test for any accessibility concerns.
One example of an automated accessibility testing engine is axe-core.
In the testing library we use a plugin that tests for detectable a11y issues on the components.

**Unit tests**
Unit tests (or DOM tests/component tests) are the tests where you test a user flow, by simulating input and querying elements.
These are the tests that require most attention.

**Testing library**
After internal discussion we decided to go for https://docs.cypress.io/. This library has many of the features we need in one library, has great documentation and is also usable for the PWA later on. More specifically we will use the 
[Cypress component testing library](https://docs.cypress.io/guides/component-testing/introduction#What-is-Component-Testing) in order to test components in isolation.

**Issues and further work**
- Add tests for components in progress (see the https://github.com/Gemeente-DenHaag/denhaag-component-library/issues/154 issue).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -107,6 +107,34 @@ This file is located here: `src/components/package.json`
...
+##### Automated a11y test
+Each component should also have an automated a11y test, to catch any detectable a11y violations, using the cypress-axe plugin:
+```tsx
+it("does not violate any accessibility rules", () => {
+  mount(<Button>Test button</Button>);
+  cy.injectAxe();
+  cy.checkA11y("button");
+});
+```
+
```
Comment (GewoonMaarten) [inline README.md:137]: Wouldn't CONTRIBUTING.md be a better place for this?

**Review by GewoonMaarten — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (DanielvanVliet) [reply README.md]: Moved it to contributing

**Review by GewoonMaarten — DISMISSED** [review]

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -200,3 +200,31 @@ Run the following command to publish a new version to the local registry (on a s
...
+
+##### Automated a11y test
+Each component should also have an automated a11y test, to catch any detectable a11y violations, using the cypress-axe plugin:
+```tsx
+it("does not violate any accessibility rules", () => {
+  mount(<Button>Test button</Button>);
+  cy.injectAxe();
+  cy.checkA11y("button");
+});
+```
```
Comment (bddjong) [inline CONTRIBUTING.md:230]: Can we add the required commands to run the tests to this README?

**Review by bddjong — CHANGES_REQUESTED** [review]
One of the button tests is failing for me. Can you reproduce this? 
image

File: .gitignore [hunk 3]
```diff
@@ -33,3 +33,5 @@ tsconfig.tsbuildinfo
 
 # story-book build output
 storybook-static
+cypress/videos
+cypress/screenshots
```
Comment (mjcarsjens) [inline .gitignore:37]: no newline :(

File: CONTRIBUTING.md (hunk 2 above)
Comment (mjcarsjens) [reply CONTRIBUTING.md:230]: +1, a simple explanation of `yarn test` and `yarn test:open` should suffice but would definitely be useful as an addition.

**Review by mjcarsjens — DISMISSED** [review]
Besides the addition to the README as requested by Brian this seems to be a fine first setup

**Comment by mjcarsjens** [comment]
> 
> 
> One of the button tests is failing for me. Can you reproduce this?
> image

@bddjong Hmm, i can't seem to reproduce this, the test pass for me when running `yarn test`.

File: CONTRIBUTING.md (hunk 2 above)
Comment (DanielvanVliet) [reply CONTRIBUTING.md:230]: Added

**Review by GewoonMaarten — APPROVED** [review]

**Review by bddjong — DISMISSED** [review]

**Review by joostvanviegen — COMMENTED** [review]
afbeelding
The test does not run for me with the yarn test command.

**Comment by DanielvanVliet** [comment]
> The tests do not run for me with the yarn test command. @DanielvanVliet

Could have been a problem with node_modules, since it works on everyone else's machines I think we can assume we will be able to fix it for you aswell.

**Review by GewoonMaarten — APPROVED** [review]

**Review by bddjong — APPROVED** [review]

**Pull request merged**